session.go

```diff
@@ -211,8 +220,8 @@ func (s *session) Retry() error {
 	if forUpdate := s.Value(forupdate.ForUpdateKey); forUpdate != nil {
 		return errors.Errorf("can not retry select for update statement")
 	}
-
 	var err error
+	s.retrying = true
```

Comment:
Move `retrying = true' to the beginning of the function and move `s.retrying = false` to defer may be better.